omnigibson/macros.py

```diff
@@ -10,33 +10,57 @@
 
 from addict import Dict
 
+
+class MacroDict(Dict):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self["_read"] = set()
+
+    def __setattr__(self, name, value):
+        if name in self.get("_read", set()):
+            raise AttributeError(f"Cannot set attribute {name} in MacroDict, it has already been used.")
+        # Use the super's setattr for setting attributes, but handle _read directly to avoid recursion.
+        if name == "_read":
```

Comment:
maybe we can have a lock / unlock feature? Only certain macros shouldn't be touched after they're read, but a lot of them can be modified later. What do you think @cgokmen ? We could take inspiration from https://github.com/ARISE-Initiative/robomimic/blob/master/robomimic/config/config.py#L37